File: src/Tests/Behavioral/GenericTypeAssertions/GenericTypeAssertions.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct Box<T>
    where T : new()
{
    public T Value;
}

public static object GetValue(nint index) {
    var values = new object[]{
        new Box<nint>(Value: 42),
        new Box<@string>(Value: "hello"u8),
        new Box<float64>(Value: 3.14F)
    }.slice();
    if (index >= 0 && index < len(values)) {
        return values[index];
    }
    return default!;
}

internal static void Main() {
    var value = GetValue(0);
    {
        var (intBox, ok) = value._<Box<nint>>(ᐧ); if (ok) {
            fmt.Printf("Found an int box with value: %d\n"u8, intBox.Value);
        }
    }
    {
        var (strBox, ok) = value._<Box<@string>>(ᐧ); if (ok){
            fmt.Printf("Found a string box with value: %s\n"u8, strBox.Value);
        } else {
            fmt.Println("Not a string box");
        }
    }
    switch (GetValue(1).type()) {
    case Box<nint> v:
        fmt.Printf("Int box: %d\n"u8, v.Value);
        break;
    case Box<@string> v:
        fmt.Printf("String box: %s\n"u8, v.Value);
        break;
    case Box<float64> v:
        fmt.Printf("Float box: %f\n"u8, v.Value);
        break;
    default: {
        var v = GetValue(1).type();
        fmt.Println("Unknown box type");
        break;
    }}
}

} // end main_package

File: src/Tests/Behavioral/GenericTypeAssertions/GenericTypeAssertions.go
package main

import "fmt"

// Box is a generic wrapper for any value
type Box[T any] struct {
	Value T
}

// GetValue is a function that returns an interface{} value
func GetValue(index int) interface{} {
	values := []interface{}{
		Box[int]{Value: 42},
		Box[string]{Value: "hello"},
		Box[float64]{Value: 3.14},
	}

	if index >= 0 && index < len(values) {
		return values[index]
	}
	return nil
}

func main() {
	// Type assertion with a generic type (TypeAssertExpr)
	value := GetValue(0)

	// Type assertion to check if the value is Box[int]
	if intBox, ok := value.(Box[int]); ok {
		fmt.Printf("Found an int box with value: %d\n", intBox.Value)
	}

	// Type assertion to check if the value is Box[string]
	if strBox, ok := value.(Box[string]); ok {
		fmt.Printf("Found a string box with value: %s\n", strBox.Value)
	} else {
		fmt.Println("Not a string box")
	}

	// Type switch with generic types
	switch v := GetValue(1).(type) {
	case Box[int]:
		fmt.Printf("Int box: %d\n", v.Value)
	case Box[string]:
		fmt.Printf("String box: %s\n", v.Value)
	case Box[float64]:
		fmt.Printf("Float box: %f\n", v.Value)
	default:
		fmt.Println("Unknown box type")
	}
}
